### PR TITLE
Allow converting NumPy datetimes to int

### PR DESCRIPTION
**Pull request by apmasell**
Add support for `int(numpy.datetime64)` and `int(numpy.timedelta64)`.

### CODE REVIEW

**Comment by stuartarchibald** [comment]
Thanks for the patch @apmasell, I think that converting `datetime64`/`timedelta64` to `int` is not permitted unless the unit is `'ns'`, whereas I think the `.view()` method should work all the time (to e.g. `.view(np.int64)`).

**Comment by sklam** [comment]
`int(timedelta64)` has different rules for the units than `int(datetime64)`:

```python
In [3]: td = np.timedelta64(1, 'Y')

In [4]: td
Out[4]: numpy.timedelta64(1,'Y')

In [5]: int(td)
Out[5]: 1
```

We'll need to check what numpy is doing to be sure.

File: numba/core/typing/builtins.py [hunk 1]
```diff
@@ -960,6 +960,13 @@ def generic(self, args, kws):
             return signature(arg, arg)
         if isinstance(arg, (types.Float, types.Boolean)):
             return signature(types.intp, arg)
+        if isinstance(arg, types.NPDatetime):
+            if arg.unit == 'ns':
+                return signature(types.int64, arg)
+            else:
+                raise errors.NumbaTypeError(f"Only datetime64[ns] can be converted, but got {arg.unit}")
```
Comment (sklam) [inline numba/core/typing/builtins.py:967]: nitpicking
```suggestion
                raise errors.NumbaTypeError(f"Only datetime64[ns] can be converted, but got datetime64[{arg.unit}]")
```

File: numba/tests/test_npdatetime.py [hunk 2]
```diff
@@ -544,6 +547,25 @@ def check(a):
...
+
+        for time in ('2014', '2016', '2000', '2014-02', '2014-03', '2014-04',
+                     '2016-02', '2000-12-31', '2014-01-16', '2014-01-05',
+                     '2014-01-07', '2014-01-06', '2014-02-02', '2014-02-27',
+                     '2014-02-16', '2014-03-01', '2000-01-01T01:02:03.002Z',
+                     '2000-01-01T01:02:03Z'):
+            check(DT(time).astype('datetime64[ns]'))
+
+        with self.assertRaises((TypeError, TypingError)):
+            f(DT('2014'))
```
Comment (sklam) [inline numba/tests/test_npdatetime.py:567]: Can you add a test when `TypingError` is raised that the expected message is shown?

**Review by sklam — APPROVED** [review]
Sorry for the delay in review. Patch looks good.

**Pull request merged**